js/modules/k6/http/request.go

```diff
@@ -73,6 +73,61 @@ func (c *Client) Request(method string, url goja.Value, args ...goja.Value) (*Re
 	return c.responseFromHTTPext(resp), nil
 }
 
+// Request makes an http request of the provided `method` and returns a corresponding response by
+// taking goja.Values as arguments
+func (c *Client) asyncRequest(method string, url goja.Value, args ...goja.Value) (*goja.Promise, error) {
+	state := c.moduleInstance.vu.State()
+	if state == nil {
+		return nil, ErrHTTPForbiddenInInitContext
+	}
+
+	var body interface{}
+	var params goja.Value
+
+	if len(args) > 0 {
+		body = args[0].Export()
+	}
+	if len(args) > 1 {
+		params = args[1]
+	}
+
+	rt := c.moduleInstance.vu.Runtime()
+	req, err := c.parseRequest(method, url, body, params)
+	if err != nil {
+		if state.Options.Throw.Bool {
+			return nil, err
```

Comment:
I must admit I'm not sure what happens at the Runtime level when you return a promise **and** an error to it. Does throwing the error take precedence over processing the promise?

This question came from me seeing anything else than a plain `goja.Promise` being returned and realizing that the code does create a promise just in time (later) and that before that, there's nothing to reject yet.

Reply:
The same as with anything else that is returned along an error - the  thing is ignored the error is thrown as exception. 

In this case this needs more tests - or more accurately needs test which differentiate between whether an exception is thrown or a promise is rejected. As *unhandled* promise rejection ends up stopping the execution either way it can be caught by the current tests as well. 

So in practice is a thing that needs changes to test that I will be doing next week.

Reply:
I see, thanks a lot for the clarification 🙇🏻 

My take away is: errors have higher precedence over promise rejections, and an exception will be thrown, before rejecting the promise is even considered 🤝

Reply:
I am not certain this is *correct* but I guess it kind of works.

As usual goja tries to map go code (and idioms) to javascript and it's idioms. Unfortunately there is no 1-to-1 mapping.

Go can have multiple returns, javascript can't. Javascript error handling is exception based, go is error/value based. 

So in the case of error being returned goja has 2 options:
1. tread it specially and convert it to exception so that it is more natural
2. return it as normal, in the case of multiple returns that is an array

So you can argue that returning an error has precedence but it is over everything as goja choose 1 which arguably is more ergonomic 99% of the cases. Even within the go ecosystem" do something wit hteh value even if you get an error" is *super* rare and I think only applise to io.Writer/Reader.  

And promises are ... just values, they are cool CS values that represent the result of an operation that can be chained with other values and operations, but still values ;).